bemani/backend/ddr/ddrsn2/ddrsn2.py

```diff
@@ -0,0 +1,566 @@
+        # Update game flags
+        for child in request.children:
+            if child.name != "flag":
+                continue
+            try:
+                value = intish(child.attribute("data"))
+                offset = intish(child.attribute("off"))
+            except ValueError:
+                continue
+
+            flags = newprofile.get_int_array("flag", 256, [1] * 256)
+            if offset < 0 or offset >= len(flags):
+                continue
+            flags[offset] = value
+            newprofile.replace_int_array("flag", 256, flags)
+
+        # Keep track of play statistics
+        self.update_play_statistics(userid, play_stats)
+
+        # Store song score data
+        for child in request.children:
+            if child.name != "music":
+                continue
+            try:
+                combo = intish(child.attribute("combo"))
+                full = intish(child.attribute("full"))
+                id = intish(child.attribute("id"))
+                mode = intish(child.attribute("mode"))
+                perf = intish(child.attribute("perf"))
+                rank = intish(child.attribute("rank"))
+                score = intish(child.attribute("score"))
+            except:
+                continue
+
+            self.update_score(
+                userid,
+                id,
+                self.game_to_db_chart(mode),
+                score * 10,
+                self.game_to_db_rank(rank),
+                self.sn2_to_db_halo(full, perf),
+                combo,
+            )
+
+        # Event Team Data
+        if game.attribute("team_i"):
+            newprofile.replace_int("team", intish(game.attribute("team_i")))
+            newprofile.replace_int(
+                "team_points", intish(game.attribute("team_p")) + newprofile.get_int("team_points", 0)
+            )
+
+        return newprofile
+
+    def handle_player_new_request(self, request: Node) -> Node:
+        refid = request.attribute("ref_id")
+        name = request.attribute("name")
+        area = intish(request.attribute("area"))
+
+        formatted_name = html.unescape(name)
+        self.new_profile_by_refid(refid, formatted_name, area)
+
+        root = Node.void("player")
+        root.add_child(Node.s8("result", 2))
+
+        return root
+
+    def handle_player_get_request(self, request: Node) -> Node:
+        refid = request.attribute("ref_id")
+        part = request.attribute("part")
+
+        root = self.get_profile_by_refid_and_part(refid, part)
+        if root is None:
+            root = Node.void("player")
+            root.set_attribute("new", "1")
+        return root
+
+    def handle_player_set_request(self, request: Node) -> Node:
+        refid = request.attribute("ref_id")
+
+        self.put_profile_by_refid(refid, request)
+
+        root = Node.void("player")
+        root.add_child(Node.s8("result", 2))
+
+        return root
+
+    def handle_info_message_request(self, request: Node) -> Node:
```

Comment:
Unsure where this API is used, figure it might get displayed to the user somehow on login but I couldn't work it out

Reply:
This could be a string-encoded server message for setting flags? Dunno, that's how bishi does things somewhat.

Reply:
Yeah the game logic seems to store the returned value to memory, but I haven't dug much further than that, I'll take a peek at the Bishi code, that might give me a few clues.